### PR TITLE
Broadcast option

### PR DESCRIPTION
**Pull request by YangFei1990**
1. Disable the horovod broadcast for every epoch
2. Add option of horovod compression

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tensorpack/train/trainers.py [hunk 1]
```diff
@@ -415,21 +441,24 @@ def _setup_graph(self, input, get_cost_fn, get_opt_fn):
...
             self.train_op = opt.apply_gradients(grads, name='train_op')
 
-        def broadcast(self):
-            logger.info("Running horovod broadcast ...")
-            # the op will be created later in initialize()
-            self.trainer._broadcast_op.run()
-
-        cb = CallbackFactory(trigger=broadcast).set_chief_only(False)
+        cb = CallbackFactory(
+            before_train=self.broadcast,
```
Comment (pidajay) [inline tensorpack/train/trainers.py]: What does this 'before_train' do? Should that not be conditioned on the BROADCAST_EVERY_EPOCH flag as well?

File: tensorpack/train/trainers.py [hunk 2]
```diff
@@ -371,37 +371,63 @@ class HorovodTrainer(SingleCostTrainer):
...
+        BROADCAST_EVERY_EPOCH (bool):
+            Whether to broadcast the variables every epoch.
+            Theoretically this is a no-op (because the variables
+            are supposed to be in-sync).
+            But this cheap operation may help prevent
+            certain numerical issues in practice.
     """
-    def __init__(self, average=True):
+
+    BROADCAST_EVERY_EPOCH = False
```
Comment (pidajay) [inline tensorpack/train/trainers.py]: Did you consider making this configurable?

Comment (YangFei1990) [reply tensorpack/train/trainers.py]: No. One reason is that this flag is inside tensorpack (There are lots of hard-coded flags in tensorpack). Also if it works fine without broadcast, we won't want to switch it back.

File: tensorpack/train/trainers.py (hunk 1 above)
Comment (YangFei1990) [reply tensorpack/train/trainers.py]: We need to run broadcast op anyway before train.

**Pull request merged**